src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,387 @@
+            (char) 11, (char) 12, (char) 14, (char) 15, (char) 16, (char) 17, (char) 18, (char) 19, (char) 20,
+            (char) 21, (char) 22, (char) 23, (char) 24, (char) 25, (char) 26, (char) 28, (char) 29, (char) 30,
+            (char) 31, (char) 127, (char) 129, (char) 141, (char) 143, (char) 144, (char) 157 };
+
+        internal static readonly UTF8Encoding utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        // take a look at the file contents and guess at the best encoding
+        internal static FileEncoding GetEncoding(string path)
+        {
+            if (!File.Exists(path))
+            {
+                return FileEncoding.Default;
+            }
+
+            byte[] initialBytes = new byte[100];
+            int bytesRead = 0;
+
+            try
+            {
+                using (FileStream stream = System.IO.File.OpenRead(path))
+                {
+                    using (BinaryReader reader = new BinaryReader(stream))
+                    {
+                        bytesRead = reader.Read(initialBytes, 0, 100);
+                    }
+                }
+            }
+            catch (IOException)
+            {
+                return FileEncoding.Default;
+            }
+
+            // Test for four-byte preambles
+            string preamble = null;
+            FileEncoding foundEncoding = FileEncoding.Default;
+
+            if (bytesRead > 3)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2], initialBytes[3]);
+
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Test for three-byte preambles
+            if (bytesRead > 2)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2]);
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Test for two-byte preambles
+            if (bytesRead > 1)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1]);
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Check for binary
+            string initialBytesAsAscii = System.Text.Encoding.ASCII.GetString(initialBytes, 0, bytesRead);
+            if (initialBytesAsAscii.IndexOfAny(nonPrintableCharacters) >= 0)
+            {
+                return FileEncoding.Byte;
+            }
+
+            // we couldn't determine anything from direct examination,
+            // return UTF8 without a BOM which should be good for both Windows and Non-Windows
+            return FileEncoding.UTF8NoBOM;
+        }
+    }
+
+}
```

Comment:
Can you check the code coverage with the newly added tests for this file? So we add tests now if we are missing any.